### PR TITLE
Binder

### PR DESCRIPTION
**Pull request by jburel**
* Add support for binder
* Add groovy script showing how to create a TrackMate model, convert tracks into omero rois.
* Add notebook matching the script (excepting the detector used).
* Split the existing page into 2: UI and API

cc @pwalczysko

### CODE REVIEW

File: docs/trackmate_ui.rst [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+Description
+-----------
+
+In this section, we show how to use TrackMate via its User Interface.
+The manual steps are essential to determine the suitable parameters to analyze the images. Note that when using the TrackMate User Interface, the generated tracks **cannot** be saved as ROIs into OMERO.server.
+
+Setup
+-----
+
+- Install Fiji on the local machine with the OMERO.insight-ij plugin. The installation instructions can be found at `here <https://omero-guides.readthedocs.io/en/latest/fiji/docs/installation.html>`_.
```
Comment (pwalczysko) [inline docs/trackmate_ui.rst]:
```suggestion
- Install Fiji on the local machine with the OMERO.insight-ij plugin. The installation instructions can be found in the `Fiji guide <https://omero-guides.readthedocs.io/en/latest/fiji/docs/installation.html>`_.
```

File: docs/trackmate_ui.rst [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+Resources
+---------
+
+-  We use an artificial track image \ https://samples.fiji.sc/FakeTracks.tif.
+
+
+Step-by-Step
+------------
+
+We will use the manual workflow to explain the steps of the script below.
```
Comment (pwalczysko) [inline docs/trackmate_ui.rst]: There is no script below here. Maybe this sentence is stale and should be removed ?

File: docs/trackmate_ui.rst [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+   screen where you can enter the name of the server to connect to,
+   the username and password. The OMERO plugin will allow you to
+   browse your data in a similar manner to OMERO.web.
+
+#. In the OMERO login dialog, click the wrench icon\ |image1| and then
+   add the server address in the dialog. By default, only "localhost"
+   is listed. Click on the *plus* icon to add a new line to the list
+   and type into the line the server address.
+
+#. Click Apply.
```
Comment (pwalczysko) [inline docs/trackmate_ui.rst]:
```suggestion
#. Click *Apply*.
```

File: docs/trackmate_ui.rst [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+   is listed. Click on the *plus* icon to add a new line to the list
+   and type into the line the server address.
+
+#. Click *Apply*.
+
+   .. image:: images/manual2.png
+
+#.  Enter your credentials and click *Login*.
+
+#. Browse the Dataset `artificial-trackmate`.
```
Comment (pwalczysko) [inline docs/trackmate_ui.rst]:
```suggestion
#. Browse the Dataset where your image to be tracked is located, e.g. `artificial-trackmate`.
```

File: docs/trackmate_ui.rst [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+#. Double-click on the `FakeTracks.tif` Image to open it in Fiji.
+
+   #. Make sure it is opened using the ``Hyperstack`` viewer. This option depends on what you did last time when using your Fiji.
+
+   #. In case you are not sure, Open ``Plugins > Bio-Formats > Bio-Formats importer`` and then select any image on your local computer to open it in Fiji. In the popup dialog, in the top left dropdown menu.
+
+      .. image:: images/Trackmate1.png
+
+   #. Select ``Hyperstack`` and click ``OK``. Then, close and re-open the FakeTracks.tif Image from OMERO.
```
Comment (pwalczysko) [inline docs/trackmate_ui.rst]:
```suggestion
   #. Select ``Hyperstack`` and click ``OK``. Then, close and re-open the `FakeTracks.tif` Image from OMERO.
```

File: docs/trackmate_ui.rst [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+      .. image:: images/Trackmate1.png
+
+   #. Select ``Hyperstack`` and click ``OK``. Then, close and re-open the `FakeTracks.tif` Image from OMERO.
+
+#. Go to ``Plugins > Tracking > TrackMate``.
+
+#. Click ``Next`` in the first dialog that pops up.
+
+#. A new dialog pop ups indicating to select a detector. Select the ``DoG detector``. Click ``Next``.
```
Comment (pwalczysko) [inline docs/trackmate_ui.rst]:
```suggestion
#. A new dialog pops up indicating to select a detector. Select the ``DoG detector``. Click ``Next``.
```

File: docs/trackmate_scripting.rst [hunk 7]
```diff
@@ -0,0 +1,127 @@
+Analyze OMERO timelapse images using the TrackMate API
+======================================================
+
+In this example we open an image stored in an OMERO server and use the `TrackMate <https://imagej.net/TrackMate>`_ API to analyze it.
+The core difference of this approach is that the generated track **can** be saved as
```
Comment (pwalczysko) [inline docs/trackmate_scripting.rst]:
```suggestion
One of the advantages of this approach over the User interface workflow is that the generated track **can** be saved as
```

File: docs/trackmate_scripting.rst [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+-  How to connect to OMERO using the JAVA API.
+
+-  How to retrieve an image.
+
+-  How to open the image using Bio-Formats.
+
+-  How to create a TrackMate model using its API.
+
+-  How to save the tracks as polygons ROIs in OMERO.
```
Comment (pwalczysko) [inline docs/trackmate_scripting.rst]:
```suggestion
-  How to save the tracks as polygon ROIs in OMERO.
```

File: docs/trackmate_scripting.rst [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+-  Script: Groovy script for tracking timelapse images - :download:`tracking.groovy <../scripts/tracking.groovy>`.
+
+Step-by-Step
+------------
+
+In this section, we go through the steps required to analyze the data.
+The script used in this document is :download:`tracking.groovy <../scripts/tracking.groovy>`.
+
+A 'Tracking' Jupyter notebook is also available in the notebooks section.
```
Comment (pwalczysko) [inline docs/trackmate_scripting.rst]: Now it is not clear where such section is and how to access it. I guess the section here is missing aka Option 2 in the ``python-Fiji`` PR.

File: docs/trackmate_scripting.rst [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+#. Launch Fiji.
+
+#. Open ``File > New > Script...``.
+
+#. Select ``Groovy`` as a language.
+
+#. Copy the content of :download:`tracking.groovy <../scripts/tracking.groovy>` in the text area.
+
+#. Enter the credentials to connect to the server and select an Image.
```
Comment (pwalczysko) [inline docs/trackmate_scripting.rst]:
```suggestion
#. A dialog will pop up. Enter the credentials to connect to the server and select an Image.
```

Comment (pwalczysko) [reply docs/trackmate_scripting.rst]: But here is a mistake in the order of appearance I guess. That dialog can only appear after you clicked Run (next step below.)

File: docs/trackmate_scripting.rst [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+#. Click Run.
+
+#. Go back to OMERO.web to visualize the tracks.
```
Comment (pwalczysko) [inline docs/trackmate_scripting.rst]:
```suggestion
#. Go back to OMERO.web to visualize the tracks. Double-click on the image in OMERO.web to open it in OMERO.iviewer.
```

**Comment by pwalczysko** [comment]
Run Fiji locally.
Used the script. Error output under details.
The console seems to indicate that login was successful. Verified the ID of the image I am passing, and this should be ok. https://outreach.openmicroscopy.org/webclient/?show=image-11600

<details>

```
Started New_.groovy at Wed Jun 17 18:44:37 BST 2020
Started New_.groovy at Wed Jun 17 18:45:03 BST 2020
omero.ClientError: Obtained null object proxy
	at omero.client.createSession(client.java:837)
	at loci.ome.io.OmeroReader.initFile(OmeroReader.java:292)
	at loci.formats.FormatReader.setId(FormatReader.java:1389)
	at loci.plugins.in.ImportProcess.initializeFile(ImportProcess.java:499)
	at loci.plugins.in.ImportProcess.execute(ImportProcess.java:142)
	at loci.plugins.in.Importer.showDialogs(Importer.java:140)
	at loci.plugins.in.Importer.run(Importer.java:76)
	at loci.plugins.LociImporter.run(LociImporter.java:78)
	at ij.IJ.runUserPlugIn(IJ.java:230)
	at ij.IJ.runPlugIn(IJ.java:193)
	at ij.IJ.runPlugIn(IJ.java:182)
	at ij.IJ$runPlugIn.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
	at Script2.open_image_plus(Script2.groovy:124)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1218)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
	at groovy.lang.DelegatingMetaClass.invokeMethod(DelegatingMetaClass.java:151)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine$2.invokeMethod(GroovyScriptEngine.java:283)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:69)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:52)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:154)
	at Script2.run(Script2.groovy:241)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:303)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:122)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:160)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:168)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:66)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:238)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

**Comment by jburel** [comment]
@pwalczysko the server was not correct in the script. This is fixed now

File: docs/trackmate_scripting.rst (hunk 9 above)
Comment (jburel) [reply docs/trackmate_scripting.rst]: I did not want to include how to run the notebook from that page.
So I will remove that part

**Comment by pwalczysko** [comment]
In notebook, built with repo2docker, on cell Create the Trackmate model.

```
Exception in thread "TrackMate spot detection thread 2/4" java.lang.OutOfMemoryError: Java heap space
	at net.imglib2.img.basictypeaccess.array.AbstractFloatArray.<init>(AbstractFloatArray.java:50)
	at net.imglib2.img.basictypeaccess.array.FloatArray.<init>(FloatArray.java:47)
	at net.imglib2.img.basictypeaccess.array.FloatArray.createArray(FloatArray.java:58)
	at net.imglib2.img.basictypeaccess.array.FloatArray.createArray(FloatArray.java:43)
	at net.imglib2.img.array.ArrayImgFactory.create(ArrayImgFactory.java:91)
	at net.imglib2.img.array.ArrayImgFactory.create(ArrayImgFactory.java:68)
	at net.imglib2.img.array.ArrayImgFactory.create(ArrayImgFactory.java:57)
	at net.imglib2.algorithm.fft2.FFT.realToComplex(FFT.java:116)
	at net.imglib2.algorithm.fft2.FFT.realToComplex(FFT.java:335)
	at net.imglib2.algorithm.fft2.FFT.realToComplex(FFT.java:74)
	at net.imglib2.algorithm.fft2.FFTConvolution.computeKernelFFT(FFTConvolution.java:578)
	at net.imglib2.algorithm.fft2.FFTConvolution.convolve(FFTConvolution.java:508)
	at fiji.plugin.trackmate.detection.LogDetector.process(LogDetector.java:145)
	at fiji.plugin.trackmate.TrackMate$1.run(TrackMate.java:385)
```

**Comment by pwalczysko** [comment]
Even with the heap space error above, the cell finished successfully creating some output.

But, finally, after saving to OMERO, the tracks are quite wild

Screen Shot 2020-06-18 at 12 36 16

**Comment by jburel** [comment]
The tracks are wild due to the detector. The one used in the script does not work in the notebook

**Comment by pwalczysko** [comment]
This https://github.com/ome/omero-guide-trackmate/pull/1#discussion_r441699562 is still persisting - some reformulation is needed I guess

**Comment by pwalczysko** [comment]
Not sure what https://github.com/ome/omero-guide-trackmate/pull/1/commits/e803b6dcbd3a1be7ad0cd1d134522f36ca48fa11 is supposed to do - I am copying and pasting the script into the Fiji scripting UI locally, as instructed - so I have to change the server address to wss  manually ? In such case the instructions need to be added.

Also, when I edit the server address with the wss, my local Fiji is answering

```
Started New_.groovy at Thu Jun 18 13:11:52 BST 2020
omero.gateway.exception.DSOutOfServiceException
	at omero.gateway.Gateway.connect(Gateway.java:304)
	at omero.gateway.Gateway$connect.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at Script1.connect_to_omero(Script1.groovy:88)
	at Script1.run(Script1.groovy:236)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:303)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:122)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:160)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:168)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:66)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:238)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: Ice.ConnectionLostException
    error = 0
	at IceInternal.AsyncResultI.__wait(AsyncResultI.java:276)
	at Ice.ObjectPrxHelperBase.end_ice_isA(ObjectPrxHelperBase.java:310)
	at Ice.ObjectPrxHelperBase.ice_isA(ObjectPrxHelperBase.java:92)
	at Ice.ObjectPrxHelperBase.ice_isA(ObjectPrxHelperBase.java:69)
	at Ice.ObjectPrxHelperBase.checkedCastImpl(ObjectPrxHelperBase.java:2810)
	at Ice.ObjectPrxHelperBase.checkedCastImpl(ObjectPrxHelperBase.java:2770)
	at Glacier2.RouterPrxHelper.checkedCast(RouterPrxHelper.java:1787)
	at omero.client.getRouter(client.java:889)
	at omero.client.createSession(client.java:810)
	at omero.gateway.Gateway.createSession(Gateway.java:1095)
	at omero.gateway.Gateway.connect(Gateway.java:293)
	... 18 more
```

**Comment by jburel** [comment]
Which version of the omero.ij plugin do you have locally?

**Comment by pwalczysko** [comment]
> Which version of the omero.ij plugin do you have locally?

Switching to 5.5.11 version of the plugin fixed the problem with the error of the script in Fiji.

I still think it should be explicitly mentioned though - the popup of the script where the server address is listed is not necessarily overtaking the default from the script, rather it takes into account what was there when you last time ran (any) Fiji script.

I would suggest https://github.com/ome/omero-guide-trackmate/pull/1/files#r442263385 to solve this, is that acceptable ?

File: docs/trackmate_scripting.rst [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+#. Launch Fiji.
+
+#. Open ``File > New > Script...``.
+
+#. Select ``Groovy`` as a language.
+
+#. Copy the content of :download:`tracking.groovy <../scripts/tracking.groovy>` in the text area.
+
+#. A dialog will pop up. Enter the credentials to connect to the server and select an Image.
```
Comment (pwalczysko) [inline docs/trackmate_scripting.rst]:
```suggestion
#. A dialog will pop up. Enter the credentials to connect to the server. Note that the server address should be in form ``wss://your-server-address/omero-ws``  or similar. This means the usage of websockets is necessary. The ending of the address (``omero-ws`` in this example) might be different for your server. 

#. Select an Image from OMERO and enter the ID of the image into the dialog in Fiji.
```

Comment (jburel) [reply docs/trackmate_scripting.rst]: The usage of websocket is not necessary.

Comment (pwalczysko) [reply docs/trackmate_scripting.rst]: But when I try it without websocket, I get  (repeatedly, 3 times)

From main console popup in Fiji

```
Parsing credentials
Logging in
-! 18/06/20 16:24:16:029 warning: SciJava-49af9ddb-Thread-0: null - createSession retry: 1
-! 18/06/20 16:24:21:658 warning: SciJava-49af9ddb-Thread-0: null - createSession retry: 2
```

From the scipt editor console (after a while) - expand details to see error

<details>

```
omero.ClientError: Obtained null object proxy
	at omero.client.createSession(client.java:837)
	at loci.ome.io.OmeroReader.initFile(OmeroReader.java:292)
	at loci.formats.FormatReader.setId(FormatReader.java:1389)
	at loci.plugins.in.ImportProcess.initializeFile(ImportProcess.java:499)
	at loci.plugins.in.ImportProcess.execute(ImportProcess.java:142)
	at loci.plugins.in.Importer.showDialogs(Importer.java:140)
	at loci.plugins.in.Importer.run(Importer.java:76)
	at loci.plugins.LociImporter.run(LociImporter.java:78)
	at ij.IJ.runUserPlugIn(IJ.java:230)
	at ij.IJ.runPlugIn(IJ.java:193)
	at ij.IJ.runPlugIn(IJ.java:182)
	at ij.IJ$runPlugIn.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
	at Script1.open_image_plus(Script1.groovy:124)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1218)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
	at groovy.lang.DelegatingMetaClass.invokeMethod(DelegatingMetaClass.java:151)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine$2.invokeMethod(GroovyScriptEngine.java:283)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:69)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:52)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:154)
	at Script1.run(Script1.groovy:241)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:303)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:122)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:160)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:168)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:66)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:238)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

Whereas with websocket it repeatedly works.

Comment (jburel) [reply docs/trackmate_scripting.rst]: you are right if it is b/c I have to specify the port for BF plugin due to https://github.com/ome/omero-blitz/issues/92

**Comment by pwalczysko** [comment]
Aha, got it - with the port specification, the script works locally.

Ready to merge fmpov

**Pull request merged**